scrypto/src/macros.rs

```diff
@@ -201,3 +201,314 @@ macro_rules! include_package {
+    };
+    (
+        $blueprint_context:tt,
+        fn $method_name:ident(self$(, $method_args:ident: $method_types:ty)*);
+        $($rest:tt)*
+    ) => {
+        compile_error!("Components cannot define methods taking self. Did you mean &self or &mut self instead?");
+    };
+    (
+        $blueprint_context:tt,
+        fn $func_name:ident($($func_args:ident: $func_types:ty),*) -> $func_output:ty;
+        $($rest:tt)*
+    ) => {
+        pub fn $func_name($($func_args: $func_types),*) -> $func_output {
+            let rtn = ::scrypto::core::Runtime::call_function(
+                ::scrypto::component::PackageAddress::from_str(::scrypto::package_address_from_context!($blueprint_context)).unwrap(),
+                ::scrypto::blueprint_name_from_context!($blueprint_context),
+                stringify!($func_name),
+                ::scrypto::args!($($func_args),*)
+            );
+            ::scrypto::buffer::scrypto_decode(&rtn).unwrap()
+        }
+        ::scrypto::bridge_blueprint_interface_members!($blueprint_context, $($rest)*);
+    };
+    (
+        $blueprint_context:tt,
+        fn $func_name:ident($($func_args:ident: $func_types:ty),*);
+        $($rest:tt)*
+    ) => {
+        pub fn $func_name($($func_args: $func_types),*) {
+            use ::scrypto::rust::str::FromStr;
+            ::scrypto::core::Runtime::call_function(
+                ::scrypto::component::PackageAddress::from_str(::scrypto::package_address_from_context!($blueprint_context)).unwrap(),
+                ::scrypto::blueprint_name_from_context!($blueprint_context),
+                stringify!($func_name),
+                ::scrypto::args!($($func_args),*)
+            );
+        }
+        ::scrypto::bridge_blueprint_interface_members!($blueprint_context, $($rest)*);
+    };
+    (
+        $blueprint_context:tt,
+    ) => {}
+}
+
+#[macro_export]
+macro_rules! package_address_from_context {
+    (
+        {
+            package: $package_address:literal,
+            blueprint: $blueprint_logical_name:literal $(,)?
+        }
+    ) => {
+        $package_address
+    };
+    () => {
+        compile_error!("Cannot call package functions (ie without &self or &mut self) on a external_component - use a external_blueprint instead.");
+    };
+    (
+        $blueprint_context:tt
+    ) => {
+        compile_error!("Unknown blueprint context - use { package: \"<PACKAGE_HEX_ADDRESS>\", blueprint: \"<BLUEPRINT_LOGICAL_NAME>\" }")
+    }
+}
+
+#[macro_export]
+macro_rules! blueprint_name_from_context {
+    (
+        {
+            package: $package_address:literal,
+            blueprint: $blueprint_logical_name:literal $(,)?
+        }
+    ) => {
+        $blueprint_logical_name
+    };
+    () => {
+        compile_error!("Cannot call package functions (ie without &self or &mut self) on a external_component - use a external_blueprint instead.");
```

Comment:
I got confused by this error message. How do we know it's used by an `external_component`?

Reply:
Add a comment explaining :)